streams/src/main/java/org/apache/kafka/streams/state/internals/ListValueStore.java

```diff
@@ -0,0 +1,147 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+import org.apache.kafka.common.utils.AbstractIterator;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.common.serialization.Serde;
+
+/**
+ * A wrapper key-value store that serializes the record values bytes as a list.
+ * As a result put calls would be interpreted as a get-append-put to the underlying RocksDB store.
+ * Range iterators would also flatten the value lists and return the values one-by-one.
+ *
+ * This store is used for cases where we do not want to de-duplicate values of the same keys but want to retain all such values.
+ */
+@SuppressWarnings("unchecked")
+public class ListValueStore
+    extends WrappedStateStore<KeyValueStore<Bytes, byte[]>, Bytes, byte[]>
+    implements KeyValueStore<Bytes, byte[]> {
+
+    static private final Serde<List<byte[]>> LIST_SERDE = Serdes.ListSerde(ArrayList.class, Serdes.ByteArray());
+
+    ListValueStore(final KeyValueStore<Bytes, byte[]> bytesStore) {
+        super(bytesStore);
+    }
+
+    @Override
+    public void put(final Bytes key, final byte[] value) {
+        // if the value is null we can skip the get and blind delete
+        if (value == null) {
+            wrapped().put(key, null);
+        } else {
+            final byte[] oldValue = wrapped().get(key);
+
+            if (oldValue == null) {
+                wrapped().put(key, LIST_SERDE.serializer().serialize(null, Collections.singletonList(value)));
+            } else {
+                final List<byte[]> list = LIST_SERDE.deserializer().deserialize(null, oldValue);
+                list.add(value);
+
+                wrapped().put(key, LIST_SERDE.serializer().serialize(null, list));
+            }
+        }
+    }
+
+    @Override
+    public byte[] putIfAbsent(final Bytes key, final byte[] value) {
+        throw new UnsupportedOperationException("putIfAbsent not supported");
+    }
+
+    @Override
+    public void putAll(final List<KeyValue<Bytes, byte[]>> entries) {
+        throw new UnsupportedOperationException("putAll not supported");
+    }
+
+    @Override
+    public byte[] delete(final Bytes key) {
+        // we intentionally disable delete calls since the returned bytes would
+        // represent a list, not a single value; we need to have a new API for delete if we do need it
+        throw new UnsupportedOperationException("delete not supported");
```

Comment:
I am fine with this, but I don't think I agree to the reasoning. `delete(k)` is just syntactic sugar for `put(k, null)`, and `get()` also returns the full list of all values... Thus, this store has some inconsistencies to a regular kv-store all over the place anyway... (what is still ok, as it's an internal store).

Reply:
Yeah maybe I'd need to make that a bit clearer: the reason I disallow caller using this call intentionally is to avoid the caller mistakenly trying to interpret/deserialize the returned byte as a single value (also note that I did not implement the deserialization of the list value in get either). Since for stream-stream join we only need the `put` and `range` calls (using `put(null)` for delete too) only, not implementing the `get/delete` is fine --- if we want to support that all the way up to the caller in KStreamImplJoin, then we need a lot more work.

Reply:
Note that in the `KStreamKStreamJoinProcessor` the shared store is still defined as `KeyValueStore<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<V1, V2>>`, hence calling its `get/delete` calls expecting to return a single `LeftOrRightValue<V1, V2>` would be doomed to fail.

Thinking about that once more, I think it's better do disable `get` as well for now.

Reply:
Seems you did not disable `get()` yet?